src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,256 @@
+﻿# Grpc C# API for error handling with Status.proto
+
+This is a protoype NuGet package providing client and server side support for the
+[gRPC richer error model](https://grpc.io/docs/guides/error/#richer-error-model).
+
+It had dependencies NuGet packages on:
+* `Google.Api.CommonProtos` - to provide the proto implementations used by the richer error model
+* `Grpc.Core.Api` - for API classes such as `RpcExcetion`
+
+Reviewers notes:
+
+* The client side borrows from ideas used in [googleapi/gax-dotnet](https://github.com/googleapis/gax-dotnet), specifically [RpcExceptionExtensions.cs](https://github.com/googleapis/gax-dotnet/blob/main/Google.Api.Gax.Grpc/RpcExceptionExtensions.cs)
+
+* The server side uses C#'s Object and Collection initializer syntax. The avoids the needs to a *builder* API to be developed.
+
+## Introduction to the richer error model
+
+Google APIs define an error model that allows a server to return an error status that is more than
+just a status code or short message.  The status returned can contain a list of various details giving
+more information about the error.
+
+This error model is define by the protocol buffers files [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+and is implemented in classes provided in the `Google.Api.CommonProtos` NuGet package.
+
+You can use this same error model in your gRPC services. This package provides classes to support you in
+doing this. In this error model the error is encapsulated by an instance of `Google.Rpc.Status` and
+returned in the trailing response metadata.  Setting and reading this metadata is handled for you. Details below.
+
+## Server Side
+
+The server side uses C#'s Object and Collection initializer syntax. The avoids the needs to a *builder* API to be developed.
+
+There are two ways that the server can return an error status:
+- by throwing an `RpcException` that contains the details of the status.
+- by explicitly set the `Status` and `ResponseTrailers` in the `ServerCallContext`
+before returning.
+
+There are examples of both methods below.  In a .NET client the result will always be
+an exception received by the client no matter which of the above methods the server
+implements.
+
+The `Google.Rpc.Status` can be created and initialized using C#'s Object and Collection initializer syntax. To add messages to the `Details` repeated field, wrap each one in `Any.Pack()` - see example below.
+
+The `Google.Rpc.Status` extension method `ToRpcException` creates the appropriate `RpcException` from the status.
+
+The `Grpc.Core.Metadata` extension method `SetRpcStatus` adds a binary representation of the status to the metadata.
+
+
+Example - throwing a RcpException:
+```C#
+throw new Google.Rpc.Status
+{
+    Code = (int)StatusCode.NotFound,
```

Comment:
needing to assign  Code = with a type cast feels odd. 

~~Idea: Since code and message can be left empty, how about having an overload `.ToRpcException(statusCode, message)` that sets the status code and message for the RpcException in a strongly typed fashion? I'm open to other ideas.~~

Reply:
on that note, what are the requirements for  Core and Message fields being set/unset and do they have to always match the Rpc's status code and message? What happens when they are different?

I think things can get confusing since we essentially have 2 sources of truth for the rpc's status code and messages so we need some clarity here.

Reply:
https://cloud.google.com/apis/design/errors#error_codes brings some more clarity to this topic, but this README.md doesn't mention the rules/recommendations for Google.Rpc.Status.Code vs Grpc.Core.Status.StatusCode at all.

My understanding is this:
- Google.Rpc.Status.Code is fully independent from Grpc.Core.Status.StatusCode and can be set to a different code and it's fine to do so.
- the recommendation is to use the values of the enum Grpc.Core.Status.StatusCode as values for Google.Rpc.Status.Code as a convention. This is a must for Google APIs and strongly recommended for third party services. But users can use a different domain of values if they want and and as long as their services are mutually compatible, things will work fine.